### PR TITLE
File 343 implement file size constraints

### PR DESCRIPTION
**Pull request by JTY009**

### CODE REVIEW

File: test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala [hunk 1]
```diff
@@ -89,6 +97,27 @@ class EnvelopeCheckerSpec extends UnitSpec {
     }
   }
 
+  "When returned envelope data has file constraint, \"10KB\"" should {
```
Comment (konradwudkowski) [inline test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala]: You can use triple quotes to avoid escaping here: `\"10KB\"`

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 2]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
                        (implicit ec: ExecutionContext) =
     EssentialAction { implicit rh =>
       Accumulator.flatten {
         check(envelopeId).map {
-          case Xor.Right("OPEN") =>
-            action(rh)
-          case Xor.Right(otherStatus) =>
-            logAndReturn(LOCKED, s"Unable to upload to envelope: $envelopeId with status: $otherStatus")
-          case Xor.Left(EnvelopeStatusNotFoundError(_)) =>
+          case (Xor.Right(envelope)) =>
```
Comment (konradwudkowski) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: parents around Xor.Right not necessary, same below for other cases

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 3]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
+            val status = (envelope \ "status").as[String]
+            if (status == "OPEN") {
+              action(checkSizeLimit(envelope))(rh)
+            } else {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: How about other status types (i.e. SEALED, DELETED etc) or envelopes that don't have status type (error handling)?

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 4]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
+  def checkSizeLimit(envelope: JsValue) = (
+    envelope \ "constraints").as[Constraints].maxSizePerItem match {
+    case Some(s) =>
+      val fileSize = s.replaceAll("[^\\d.]", "").toInt + 1
+      val fileSizeType = s.toUpperCase.replaceAll("[^KB,MB]{2}", "")
+      fileSizeType match {
+        case "KB" => fileSize * 1024
+        case "MB" => fileSize * 1024 * 1024
+      }
+    case None => 11 * 1024 * 1024
```
Comment (konradwudkowski) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: if we provide a default of 11mb it might be better to be more explicit about it and keep it in a variable. Also when I read this method and its name `check` I assumed return type to be

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 5]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
-          case Xor.Left(error) =>
+          case (Xor.Left(error)) =>
             logAndReturn(INTERNAL_SERVER_ERROR, error.toString)
         }
       }
     }
 
+  def checkSizeLimit(envelope: JsValue) = (
+    envelope \ "constraints").as[Constraints].maxSizePerItem match {
+    case Some(s) =>
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: use logical variable names at all times please

File: app/uk/gov/hmrc/fileupload/controllers/FileUploadController.scala [hunk 6]
```diff
@@ -30,27 +30,25 @@ import uk.gov.hmrc.fileupload.controllers.FileUploadController._
...
-
   def uploadWithEnvelopeValidation(envelopeId: EnvelopeId, fileId: FileId) =
     withValidEnvelope(envelopeId) {
-      upload(envelopeId, fileId)
+      limit => upload(limit)(envelopeId, fileId)
     }
 
-  def upload(envelopeId: EnvelopeId, fileId: FileId) =
-    Action.async(parse.maxLength(MAX_FILE_SIZE_IN_BYTES, uploadParser())) { implicit request =>
+  def upload(limit: Int)(envelopeId: EnvelopeId, fileId: FileId) = {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/FileUploadController.scala]: perhaps rename to fileSizeLimit to make it more explanatory

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 7]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
             logAndReturn(NOT_FOUND, s"Unable to upload to nonexistent envelope: $envelopeId")
-          case Xor.Left(error) =>
+          case (Xor.Left(error)) =>
             logAndReturn(INTERNAL_SERVER_ERROR, error.toString)
         }
       }
     }
 
+  def checkSizeLimit(envelope: JsValue) = (
+    envelope \ "constraints").as[Constraints].maxSizePerItem match {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: How about the envelopes that don't have constraints field? i.e. existing envelopes in Mongodb?

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 8]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
+    case Some(s) =>
+      val fileSize = s.replaceAll("[^\\d.]", "").toInt + 1
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: why increase file size by 1 here?

Comment (JTY009) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: Because its the form plus the file

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 9]
```diff
@@ -20,46 +20,61 @@ import cats.data.Xor
...
+      val fileSizeType = s.toUpperCase.replaceAll("[^KB,MB]{2}", "")
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: `fileSizeType` returns the original attribute value i.e. _2MB_ ? Have you actually tested this functionality?

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala (hunk 8 above)
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: form? So we assume form's size is 1MB if the user specified 5MB as the constrain attribute?

File: test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala [hunk 10]
```diff
@@ -89,6 +97,27 @@ class EnvelopeCheckerSpec extends UnitSpec {
...
+    "set the as upload size limit to 11KB" in {
+      val envelope10KB = Json.parse("""{"status" : "OPEN", "constraints": { "maxSizePerItem" : "10KB" } }""")
```
Comment (KorhanOzturk90) [inline test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala]: Probably worth adding a test to cover constraints added with 'MB'

**Comment by JTY009** [comment]
Up for Review

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 11]
```diff
@@ -20,46 +20,90 @@ import cats.data.Xor
...
+    val definedConstraints = (envelope \ "constraints").asOpt[Constraints]
+       definedConstraints match {
+           //TODO Possibly in the future this may change to take into account the envelope size e.g. if envelope size is 5MB and the upload limit is 10MB
+         case Some(constraints) => constraints.maxSizePerItem match {
+           case Some(maxSizePerItem) =>
+             val fileSize = maxSizePerItem.replaceAll("[^\\d.]", "").toInt
+             val fileSizeType = maxSizePerItem.toUpperCase.replaceAll("[^KMB]", "")
+             fileSizeType match {
+               case "KB" =>
+                 if(fileSize <= 1024) {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: you might want to use [guard expressions](https://www.safaribooksonline.com/library/view/scala-cookbook/9781449340292/ch03s14.html) here to replace those if else statements to improve readability.

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 12]
```diff
@@ -20,46 +20,90 @@ import cats.data.Xor
...
+                   fileSize * 1024
+                 } else {
+                   //If maxSizePerItem is accidentally more than 10MB e.g. 10000KB
+                   defaultFileSize
+                 }
+               case "MB" =>
+                 if(fileSize <= 10){
+                   fileSize * 1024 * 1024
+                 } else {
+                   //If maxSizePerItem is accidentally more than 10MB e.g. 100MB
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: Personally not a fan of line comments in the code. Shouldn't be needed when the code is readable enough.

**Comment by KorhanOzturk90** [comment]
Please see my comments from yesterday about the use of guards expression and removal of comments @JTY009

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 13]
```diff
@@ -20,46 +20,90 @@ import cats.data.Xor
...
-        check(envelopeId).map {
-          case Xor.Right("OPEN") =>
-            action(rh)
-          case Xor.Right(otherStatus) =>
-            logAndReturn(LOCKED, s"Unable to upload to envelope: $envelopeId with status: $otherStatus")
-          case Xor.Left(EnvelopeStatusNotFoundError(_)) =>
+        checkEnvelopeDetails(envelopeId).map {
+          case Xor.Right(envelope) =>
+            val status = (envelope \ "status").as[String]
+            if (status == "OPEN") {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: Probably better to convert to pattern matching instead of if-else statements on envelope.status

File: app/uk/gov/hmrc/fileupload/quarantine/events.scala [hunk 14]
```diff
@@ -20,7 +20,7 @@ import play.api.libs.json.{Format, JsObject, Json}
 import uk.gov.hmrc.fileupload.{EnvelopeId, Event, FileId, FileRefId}
 
 case class FileInQuarantineStored(envelopeId: EnvelopeId, fileId: FileId, fileRefId: FileRefId,
-                                  created: Long, name: String, contentType: String, metadata: JsObject) extends Event
+                                  created: Long, name: String, fileLength: Long, contentType: String, metadata: JsObject) extends Event
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/quarantine/events.scala]: What's the purpose of adding this new field here?

File: test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala [hunk 15]
```diff
@@ -38,13 +39,18 @@ class EnvelopeCheckerSpec extends UnitSpec {
...
   val testEnvelopeId = EnvelopeId()
 
+  val defaultFileSize = 10 * 1024 * 1024
+
   "When an envelope is OPEN it" should {
     "be possible to execute an Action" in {
       val expectedAction = Action { req => Ok }
 
-      val wrappedAction = withValidEnvelope(_ => Future(Xor.right("OPEN")))(testEnvelopeId)(expectedAction)
+      val envelopeOpen = Json.parse("""{ "status" : "OPEN", "constraints" : {} }""")
```
Comment (KorhanOzturk90) [inline test/uk/gov/hmrc/fileupload/controllers/EnvelopeCheckerSpec.scala]: It'd be nice to add a test case with an envelope that has no constraint attribute at all in addition to empty constraints field.

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala [hunk 16]
```diff
@@ -20,46 +20,90 @@ import cats.data.Xor
...
+            }
+          case Xor.Left(EnvelopeDetailNotFoundError(_)) =>
             logAndReturn(NOT_FOUND, s"Unable to upload to nonexistent envelope: $envelopeId")
           case Xor.Left(error) =>
             logAndReturn(INTERNAL_SERVER_ERROR, error.toString)
         }
       }
     }
 
+  def setMaxFileSize(envelope: JsValue): FileSize = {
```
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: wrong file naming, this method retrieves allowed max file size limit from envelope Json, does not set it.

**Comment by KorhanOzturk90** [comment]
Looking good. Added couple more commands around naming/testing.

File: app/uk/gov/hmrc/fileupload/quarantine/events.scala (hunk 14 above)
Comment (YuanZHANG910) [inline app/uk/gov/hmrc/fileupload/quarantine/events.scala]: this fileLength will be used at the backend, the backend does the finial checks when the file is in quarantine, those checks need to know the fileLength of a file and that is come from here (if am right).

File: app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala (hunk 12 above)
Comment (JTY009) [inline app/uk/gov/hmrc/fileupload/controllers/EnvelopeChecker.scala]: I only written it there to avoid confusion.

File: app/uk/gov/hmrc/fileupload/quarantine/events.scala (hunk 14 above)
Comment (KorhanOzturk90) [inline app/uk/gov/hmrc/fileupload/quarantine/events.scala]: OK that makes sense (even though that addition should have been made in a different branch)

**Comment by KorhanOzturk90** [comment]
Looking good to me & ready to merge

**Comment by kastoestoramadus** [comment]
On slack was said there is some regexp bug. @YuanZHANG910 When it will be fixed?

**Comment by YuanZHANG910** [comment]
I have no idea how the bug happened, i signed off from those tickets about 2 weeks ago :(

**Comment by KorhanOzturk90** [comment]
Not entirely sure which bug you are referring to @kastoestoramadus but there is no more any regex expression within the changes that been made by this PR so it should be fixed
